File: frontend/src/types/opencti.ts
// Mirrors backend/app/connectors/opencti/schema/opencti.py. The backend already
// flattens OpenCTI's GraphQL shapes (edges/node, objectLabel, createdBy{name}),
// so these are plain records.

export interface OpenCTILabel {
	value: string
	color: string | null
}

export interface OpenCTIObjectMeta {
	id: string
	standard_id: string | null
	entity_type: string
	created_at: string | null
	updated_at: string | null
	created_by: string | null
	labels: OpenCTILabel[]
	/** Marking definitions, e.g. "TLP:CLEAR" */
	markings: string[]
}

export interface OpenCTIIndicator extends OpenCTIObjectMeta {
	name: string | null
	description: string | null
	pattern: string | null
	pattern_type: string | null
	main_observable_type: string | null
	score: number | null
	confidence: number | null
	valid_from: string | null
	valid_until: string | null
	revoked: boolean | null
}

export interface OpenCTIHash {
	algorithm: string
	hash: string
}

export interface OpenCTIReportRef {
	id: string
	name: string | null
	published: string | null
}

export interface OpenCTIObservable extends OpenCTIObjectMeta {
	value: string | null
	description: string | null
	score: number | null
	file_name: string | null
	hashes: OpenCTIHash[]
	indicators: OpenCTIIndicator[]
	indicators_count: number
	reports: OpenCTIReportRef[]
	reports_count: number
}

export interface OpenCTIExternalReference {
	source_name: string | null
	url: string | null
	external_id: string | null
	description: string | null
}

export interface OpenCTIEntity extends OpenCTIObjectMeta {
	parent_types: string[]
	name: string | null
	description: string | null
	confidence: number | null
	score: number | null
	external_references: OpenCTIExternalReference[]
}

export interface OpenCTIAbout {
	version: string | null
	dependencies: { name: string; version: string | null }[]
	user_name: string | null
	user_email: string | null
}

export interface OpenCTIPageInfo {
	global_count: number | null
	has_next_page: boolean
	end_cursor: string | null
}

export interface OpenCTIAvailability {
	configured: boolean
	verified: boolean
	/** OpenCTI's web address, only set once verified. Links use `${platform_url}/dashboard/id/${id}`. */
	platform_url: string | null
}

export interface OpenCTIObservableLookup {
	value: string
	found: boolean
	total: number
	observables: OpenCTIObservable[]
}

export interface OpenCTIValueLookup {
	value: string
	found: boolean
	observables: OpenCTIObservable[]
}

export interface OpenCTIBatchLookup {
	/** One entry per requested value, in request order. */
	results: OpenCTIValueLookup[]
	/** OpenCTI matched more than one query returns; a "not found" may be a false negative. */
	truncated: boolean
}

export interface OpenCTIIndicatorsQuery {
	search?: string
	first?: number
	after?: string
	min_score?: number
	main_observable_type?: string
}

File: backend/app/connectors/opencti/schema/opencti.py
from typing import List
from typing import Optional

from pydantic import BaseModel
from pydantic import Field

# OpenCTI answers in nested Relay/GraphQL shapes (`edges { node }`,
# `objectLabel`, `createdBy { name }`, …). The services flatten them into the
# models below so CoPilot callers never have to know that shape.


class OpenCTILabel(BaseModel):
    value: str
    color: Optional[str] = None


class OpenCTIPageInfo(BaseModel):
    global_count: Optional[int] = Field(default=None, description="Total number of matches on the server")
    has_next_page: bool = False
    end_cursor: Optional[str] = Field(default=None, description="Pass as `after` to fetch the next page")


class OpenCTIObjectMeta(BaseModel):
    id: str = Field(..., description="OpenCTI internal id")
    standard_id: Optional[str] = Field(default=None, description="STIX id, e.g. indicator--…")
    entity_type: str
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    created_by: Optional[str] = Field(default=None, description="Name of the author identity")
    labels: List[OpenCTILabel] = Field(default_factory=list)
    markings: List[str] = Field(default_factory=list, description="Marking definitions, e.g. TLP:CLEAR")


class OpenCTIIndicator(OpenCTIObjectMeta):
    name: Optional[str] = None
    description: Optional[str] = None
    pattern: Optional[str] = None
    pattern_type: Optional[str] = None
    main_observable_type: Optional[str] = None
    score: Optional[int] = None
    confidence: Optional[int] = None
    valid_from: Optional[str] = None
    valid_until: Optional[str] = None
    revoked: Optional[bool] = None


class OpenCTIHash(BaseModel):
    algorithm: str
    hash: str


class OpenCTIReportRef(BaseModel):
    id: str
    name: Optional[str] = None
    published: Optional[str] = None


class OpenCTIObservable(OpenCTIObjectMeta):
    value: Optional[str] = None
    description: Optional[str] = None
    score: Optional[int] = None
    file_name: Optional[str] = None
    hashes: List[OpenCTIHash] = Field(default_factory=list)
    indicators: List[OpenCTIIndicator] = Field(default_factory=list)
    indicators_count: int = Field(default=0, description="Total indicators based on this observable")
    reports: List[OpenCTIReportRef] = Field(default_factory=list)
    reports_count: int = Field(default=0, description="Total reports that contain this observable")


class OpenCTIExternalReference(BaseModel):
    source_name: Optional[str] = None
    url: Optional[str] = None
    external_id: Optional[str] = None
    description: Optional[str] = None


class OpenCTIEntity(OpenCTIObjectMeta):
    parent_types: List[str] = Field(default_factory=list)
    name: Optional[str] = Field(default=None, description="OpenCTI's representative name for the object")
    description: Optional[str] = None
    confidence: Optional[int] = None
    score: Optional[int] = None
    external_references: List[OpenCTIExternalReference] = Field(default_factory=list)


class OpenCTIDependency(BaseModel):
    name: str
    version: Optional[str] = None


class OpenCTIAbout(BaseModel):
    version: Optional[str] = None
    dependencies: List[OpenCTIDependency] = Field(default_factory=list)
    user_name: Optional[str] = Field(default=None, description="The account the connector token belongs to")
    user_email: Optional[str] = None


# ── Responses ────────────────────────────────────────────────────────────────


class OpenCTIAvailabilityResponse(BaseModel):
    success: bool
    message: str
    configured: bool = Field(..., description="The connector row exists with a URL and an API key")
    verified: bool = Field(..., description="The last Verify on the Connectors page succeeded")
    platform_url: Optional[str] = Field(
        default=None,
        description="OpenCTI's web address, for 'open in OpenCTI' links (<platform_url>/dashboard/id/<id>). Only set once verified.",
    )


class OpenCTIAboutResponse(BaseModel):
    success: bool
    message: str
    about: OpenCTIAbout


class OpenCTIObservableLookupResponse(BaseModel):
    success: bool
    message: str
    value: str = Field(..., description="The value that was looked up")
    found: bool = Field(..., description="True when OpenCTI knows at least one matching observable")
    total: int = 0
    observables: List[OpenCTIObservable] = Field(default_factory=list)


class OpenCTIIndicatorsResponse(BaseModel):
    success: bool
    message: str
    indicators: List[OpenCTIIndicator] = Field(default_factory=list)
    page_info: OpenCTIPageInfo


class OpenCTIEntityResponse(BaseModel):
    success: bool
    message: str
    entity: OpenCTIEntity
